spring-cloud-azure-keyvault-config/src/main/java/com/microsoft/azure/spring/cloud/keyvault/config/KeyVaultPropertySource.java

```diff
@@ -6,24 +6,71 @@
 
 package com.microsoft.azure.spring.cloud.keyvault.config;
 
+import com.microsoft.azure.PagedList;
+import com.microsoft.azure.keyvault.KeyVaultClient;
+import com.microsoft.azure.keyvault.models.SecretItem;
 import org.springframework.core.env.EnumerablePropertySource;
+import org.springframework.util.StringUtils;
+
+import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * Retrieve all properties from the Azure Key Vault instance.
  */
-// TODO: Use Key Vault source class here.
-public class KeyVaultPropertySource extends EnumerablePropertySource<Object> {
-    public KeyVaultPropertySource(String name, Object source) {
-        super(name, source);
+public class KeyVaultPropertySource extends EnumerablePropertySource<KeyVaultClient> {
+
+    private final KeyVaultClient keyVaultClient;
+    private final String vaultUrl;
+
+    // Saves names of all secrets in a Key Vault. Key is secret name. Value is null.
+    private final ConcurrentHashMap<String, String> secretNames = new ConcurrentHashMap<>();
+    // Saves values of each secret. Key is secret name. Value is secret value.
+    private final ConcurrentHashMap<String, String> secretValues = new ConcurrentHashMap<>();
+
+    public KeyVaultPropertySource(String name, KeyVaultClient keyVaultClient, String vaultUrl) {
+        super(name, keyVaultClient);
+        this.keyVaultClient = keyVaultClient;
+        this.vaultUrl = vaultUrl;
+
+        listSecretsFromKeyVault();
     }
 
     @Override
     public String[] getPropertyNames() {
-        return new String[0];
+        return this.secretNames.keySet().toArray(new String[this.secretValues.size()]);
     }
 
     @Override
-    public Object getProperty(String s) {
-        return null;
+    public Object getProperty(String name) {
+        if (StringUtils.isEmpty(name)) {
+            return null;
+        }
+
+        // NOTE:
+        // Azure Key Vault secret name pattern is: ^[0-9a-zA-Z-]+$
+        // "." is not allowed.
+        final String normalizedName = name.replace(".", "-");
+
+        if (this.secretNames.containsKey(normalizedName)) {
+            return getSecretFromKeyVault(normalizedName);
+        } else {
+            return null;
+        }
+    }
+
+    private void listSecretsFromKeyVault() {
+        final PagedList<SecretItem> secrets = this.keyVaultClient.listSecrets(this.vaultUrl);
+        secrets.loadAll();
+
+        final String vaultSecretsBaseUrl = this.vaultUrl + "/secrets/";
+        secrets.stream()
+                .map(s -> s.id().replace(vaultSecretsBaseUrl, ""))

```

Comment:
How about use substring of last index of slash? This will improve performance since current implementation need to match the long full string.